### ISSUE TITLE
Perceptual precision comparison does not appear to work correctly

### ISSUE THREAD

**Issue by jaanus** [pointfreeco/swift-snapshot-testing#673]
**Describe the bug**

Testing with perceptual snapshot comparison fails because the perceptual precision computation returns a value that does not match my perception.

**To Reproduce**

```swift
struct ContentView: View {
  var body: some View {
    VStack {
      Image(systemName: "globe")
        .imageScale(.large)
        .foregroundColor(.accentColor)
      Text("Hello, world!")
    }
    .padding()
  }
}

final class PerceptualSnapshotBugTests: XCTestCase {
  func test_view() throws {
    let sut = ContentView()
    assertSnapshot(matching: sut, as: .image(perceptualPrecision: 0.5))
  }
}
```

Record a snapshot and then modify it very slightly with an image editing tool.

**Expected behavior**

The above test should pass, since the perceptual difference between reference and test result is small.

**Screenshots**

The above test fails. When I debug to the perceptual comparison code, I see that `actualPixelPrecision` makes sense, but `actualPerceptualPrecision` is reported much lower than expected.

Here is screenshot with Xcode and the reference image. You see that I modified the reference image by slightly cutting away from the “H” character.

<img width="1675" alt="Screenshot 2022-11-02 at 15 46 33" src="https://user-images.githubusercontent.com/184253/199508047-579d21f4-4663-414c-b1be-a57d906b0330.png">

**Environment**

  - swift-snapshot-testing version 1.10.0
  - Xcode 14.1 (14B47b)
  - Swift: whatever is bundled with that Xcode
  - OS: macOS Ventura 13.0 (22A380)

**Comment by cooksimo** [comment]
I'm having similar issues, any thoughts @ejensen?

**Comment by iOSBrett** [comment]
Any update on this, I am seeing issues also. Images are identical (difference map has every single pixel as 0, 0, 0) and it fails with:
Actual perceptual precision 0.82234377 is less than required 0.98

![difference_3_21598FE1-C511-4829-B7D6-D3EB8DD2EA3E](https://user-images.githubusercontent.com/436052/213636215-0fd6b614-ad37-48f5-a018-39b3df82d5b9.png)

**Comment by lukeredpath** [comment]
I'm also having the same problem. No apparent difference but the perceptual precision is much lower.

**Comment by ZevEisenberg** [comment]
FWIW Kaleidoscope 3 shows that these pixels are different in some way:
<img width="415" alt="image" src="https://github.com/pointfreeco/swift-snapshot-testing/assets/464574/0367e25b-3053-41e3-aa0a-384fa134a28d">

**Comment by ZevEisenberg** [comment]
🤔 I think the difference that actually matters here is that one rando pixel on the left side of the image, but kinda near the middle. It will look like dirt on your screen, so zoom in and scroll and you'll see it move around. You can clearly the the difference in the image. Not sure what would cause it, though.

Extra pixel:

<img width="86" alt="image" src="https://github.com/pointfreeco/swift-snapshot-testing/assets/464574/32ee17ca-c9a5-456a-9d8c-800787967fbf">

No extra pixel:

<img width="127" alt="image" src="https://github.com/pointfreeco/swift-snapshot-testing/assets/464574/5c2e2ef3-fe02-4289-945e-1dcccaa7b339">

**Comment by CraigSiemens** [comment]
Part of the issue seems to be related to how `CILabDeltaE` handles transparency.

The two images seem to have slightly different shadow colors and opacities
`#00172E` `#0B202B` which leads to `CILabDeltaE` reporting them ad fairly different.

If the two images are changed so they're on a solid background, the resulting colors become nearly identical, so the required `perceptualPrecision` goes from `~0.93` to `~0.99`.

<img width='25%' src='https://github.com/pointfreeco/swift-snapshot-testing/assets/680442/4629017c-4f1f-4733-8720-7e7f98bd4548'>
<img width='25%' src='https://github.com/pointfreeco/swift-snapshot-testing/assets/680442/b4ecb9fb-6413-491a-bd3e-00905ed45884'>

---

The other part of the issue seems to be around the messaging when it doesnt match
```
Actual perceptual precision 0.43562502 is less than required 0.98
```
The value printed out here is `(1 - delta e of the worst pixel) / 100` in the image, in this case the pixel that's blue in one image and what in the other as mentioned above. This is printing out a value that makes the accuracy of the snapshot look far worse than it actually is. By comparison, the average across the whole image would be more like `0.9949010414` which isn't really that bad.

I'd suggest removing that line entirely since it doesn't provide any actionable information. The value is the `perceptualPrecision` that would need to be used to make it pass only if it had a `pixelPrecision` of 1.0. Since `perceptualPrecision` is used on a per pixel basis, I'm not sure there's a number that could be printed out that would be useful.

**Comment by mbrandonw** [comment]
@babbage Please do not paste a wall of AI text into discussions. It is not helpful at all and only detracts from the work others have done in this thread to research the problem. In the future it would be better to summarize in your own words (seems like just a few sentences would do it) or open a PR if you wish (no promises it will be merged).

**Comment by babbage** [comment]
Fair enough, apologies. I've since identified that #665 already fixes this (for our failure case), by resolving the colourspace mismatches that are created by SnapshotTesting's internal comparison pipeline, so most of what I posted was re-treading ground. A key symptom for others to look out for is intermittent failures that are due to the saturation level of the colours in the specific content of particular snapshots, when captured in a P3 colourspace like iOS. A cause that can be hard to identify. So would be great if that PR was merged. It predates all of the discussion here.

**Comment by hristosvilensky-ship-it** [comment]
Another data point, on 1.19.4 (Xcode 27, iOS 27.0 simulator, 16-bit Display P3 references, 1206 x 2622):

Reference and render differ in exactly **2** pixels out of 3,162,132, in one channel only (B, +288 of 65535). Decoding both PNGs at full 16-bit depth gives a worst-pixel deltaE of **0.545**. The library reports:

```
The percentage of pixels that match 0.87701416 is less than required 0.99
The lowest perceptual color precision 0.95097655 is less than required 0.98
```

i.e. 12.3% of pixels above deltaE 2 and a worst-pixel deltaE of 4.90. A second capture, 4 differing pixels with a real worst deltaE of 0.678, reports `0.94247437 / 0.9423438` (worst deltaE 5.77).

Those numbers repeat identically to 8 decimals across two days and two simulator clones, so this is deterministic rather than flaky — which fits the colour-space diagnosis above: `perceptuallyCompare` receives `CIImage(cgImage:)` of the un-normalised 16-bit P3 images, while the byte path compares through the 8-bit sRGB `context(for:)`.

Practical consequence for 16-bit references: once a single pixel differs, the perceptual tolerance saves nothing. The workaround we settled on is `perceptualPrecision: 1` (which takes the CPU byte-count branch in the `else` of `if perceptualPrecision < 1`) with `precision` derived from a byte budget instead of a round fraction.
